core/src/main/java/org/wso2/msf4j/internal/websocket/EndpointsRegistryImpl.java

```diff
@@ -0,0 +1,146 @@
+ *  under the License.
+ */
+
+package org.wso2.msf4j.internal.websocket;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.wso2.carbon.messaging.CarbonMessage;
+import org.wso2.carbon.transport.http.netty.common.Constants;
+import org.wso2.msf4j.internal.router.PatternPathRouter;
+import org.wso2.msf4j.websocket.WebSocketEndpointsRegistry;
+import org.wso2.msf4j.websocket.exception.WebSocketEndpointAnnotationException;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+/**
+ * Implementation for {@link WebSocketEndpointsRegistry}.
+ * Endpoints are Dispatched and Stored in a {@link PatternPathRouter}. So when new request comes it will be routed to
+ * best matching URI.
+ */
+public class EndpointsRegistryImpl implements WebSocketEndpointsRegistry {
+
+    private static final Logger log = LoggerFactory.getLogger(EndpointsRegistryImpl.class);
+    private static final EndpointsRegistryImpl webSocketEndpointsRegistry = new EndpointsRegistryImpl();
+
+    // Map <uri, WebSocketEndpoint>
+    private final Map<String, Object> webSocketEndpointMap = new ConcurrentHashMap<>();
+    // PatterPathRouter<WebSocketEndpoint>
+    private PatternPathRouter<Object> endpointPatternPathRouter = PatternPathRouter.create();
+
+    // Makes this class singleton.
+    private EndpointsRegistryImpl() {
+    }
+
+    /**
+     * @return this singleton instance of {@link EndpointsRegistryImpl}.
+     */
+    public static EndpointsRegistryImpl getInstance() {
+        return webSocketEndpointsRegistry;
+    }
+
+    /**
+     * Adding endpoints to the registry.
+     * @param webSocketEndpoints to add.
+     */
+    public void addEndpoint(Object... webSocketEndpoints) throws WebSocketEndpointAnnotationException {
+        for (Object endpoint: webSocketEndpoints) {
+            EndpointDispatcher dispatcher = new EndpointDispatcher();
+            webSocketEndpointMap.put(dispatcher.getUri(endpoint), endpoint);
+            updatePatternPathRouter();
+            log.info("Endpoint Registered : " + dispatcher.getUri(endpoint));
+        }
+    }
+
+    /**
+     * Remove WebSocket Endpoint from Registry.
+     * @param webSocketEndpoint which should be removed.
+     * @throws WebSocketEndpointAnnotationException throws when WebSocket {@link javax.websocket.server.ServerEndpoint}
+     * is no declared in the endpoint.
+     */
+    public void removeEndpoint(Object webSocketEndpoint) throws WebSocketEndpointAnnotationException {
+        EndpointDispatcher dispatcher = new EndpointDispatcher();
+        webSocketEndpointMap.remove(dispatcher.getUri(webSocketEndpoint));
+        updatePatternPathRouter();
+        log.info("Removed endpoint : " + dispatcher.getUri(webSocketEndpoint));
+    }
+
+    /**
+     * Return the best possible {@link org.wso2.msf4j.internal.router.PatternPathRouter.RoutableDestination}.
```

Comment:
Include a blank line after the method description

eg: /* Description
       *
       *@param ...